packages/api/src/router/widgets/downloads.ts

```diff
@@ -0,0 +1,135 @@
+            unsubscribe();
+          });
+        };
+      });
+    }),
+  pause: publicProcedure
+    .unstable_concat(
+      createManyIntegrationMiddleware("interact", "sabNzbd", "nzbGet", "qBittorrent", "deluge", "transmission"),
+    )
+    .mutation(async ({ ctx }) => {
+      await Promise.all(
+        ctx.integrations.map(async (integration) => {
+          const integrationInstance = getIntegrationInstance(integration.kind, integration);
+          await integrationInstance.pauseQueueAsync();
+        }),
+      );
+    }),
+  pauseItem: publicProcedure
+    .unstable_concat(
+      createManyIntegrationMiddleware("interact", "sabNzbd", "nzbGet", "qBittorrent", "deluge", "transmission"),
+    )
+    .input(z.object({ item: z.any() satisfies z.ZodType<DownloadClientItem> }))
+    .mutation(async ({ ctx, input }) => {
+      await Promise.all(
+        ctx.integrations.map(async (integration) => {
+          const integrationInstance = getIntegrationInstance(integration.kind, integration);
+          await integrationInstance.pauseItemAsync(input.item as DownloadClientItem);
+        }),
+      );
+    }),
+  resume: publicProcedure
+    .unstable_concat(
+      createManyIntegrationMiddleware("interact", "sabNzbd", "nzbGet", "qBittorrent", "deluge", "transmission"),
+    )
+    .mutation(async ({ ctx }) => {
+      await Promise.all(
+        ctx.integrations.map(async (integration) => {
+          const integrationInstance = getIntegrationInstance(integration.kind, integration);
+          await integrationInstance.resumeQueueAsync();
+        }),
+      );
+    }),
+  resumeItem: publicProcedure
+    .unstable_concat(
+      createManyIntegrationMiddleware("interact", "sabNzbd", "nzbGet", "qBittorrent", "deluge", "transmission"),
+    )
+    .input(z.object({ item: z.any() satisfies z.ZodType<DownloadClientItem> }))
+    .mutation(async ({ ctx, input }) => {
+      await Promise.all(
+        ctx.integrations.map(async (integration) => {
+          const integrationInstance = getIntegrationInstance(integration.kind, integration);
+          await integrationInstance.resumeItemAsync(input.item as DownloadClientItem);
+        }),
+      );
+    }),
+  deleteItem: publicProcedure
+    .unstable_concat(
+      createManyIntegrationMiddleware("interact", "sabNzbd", "nzbGet", "qBittorrent", "deluge", "transmission"),
+    )
+    .input(z.object({ item: z.any() satisfies z.ZodType<DownloadClientItem>, fromDisk: z.boolean() }))
+    .mutation(async ({ ctx, input }) => {
+      await Promise.all(
+        ctx.integrations.map(async (integration) => {
+          const integrationInstance = getIntegrationInstance(integration.kind, integration);
+          await integrationInstance.deleteItemAsync(input.item as DownloadClientItem, input.fromDisk);
+        }),
+      );
+    }),
+});
+
+function getIntegrationInstance(kind: IntegrationKind, integration: IntegrationInput): DownloadClientIntegration {
```

Comment:
Instead of writing your own abstraction you could also replace the current `integrationCreatorByKind` method with the following implementation:

```ts
export const integrationCreatorByKind = <TKind extends keyof typeof integrationCreators>(
  kind: TKind,
  integration: IntegrationInput,
) => {
  if (!(kind in integrationCreators)) {
    throw new Error(`Unknown integration kind ${kind}. Did you forget to add it to the integration creator?`);
  }

  return new integrationCreators[kind](integration) as InstanceType<(typeof integrationCreators)[TKind]>;
};

const integrationCreators = {
  piHole: PiHoleIntegration,
  adGuardHome: AdGuardHomeIntegration,
  homeAssistant: HomeAssistantIntegration,
  jellyfin: JellyfinIntegration,
  sonarr: SonarrIntegration,
  sabNzbd: SabnzbdIntegration,
  nzbGet: NzbGetIntegration,
  qBittorrent: QBitTorrentIntegration,
  deluge: DelugeIntegration,
  transmission: TransmissionIntegration,
} satisfies Partial<Record<IntegrationKind, new (integration: IntegrationInput) => Integration>>;
```

This one will be typesafe regarding input kind and return instances

Reply:
I would never have came up with that, Thanks a whole lot